src/sources/mqtt/source.rs

```diff
@@ -0,0 +1,125 @@
+use snafu::Snafu;
+use vector_lib::tls::TlsError;
+
+use super::config::ConfigurationError;
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub))]
+pub enum MqttError {
+    #[snafu(display("TLS error: {}", source))]
+    Tls { source: TlsError },
+    #[snafu(display("MQTT configuration error: {}", source))]
+    Configuration { source: ConfigurationError },
+}
+
+#[derive(Clone)]
+pub struct MqttConnector {
+    options: MqttOptions,
+    topic: String,
+}
+
+impl MqttConnector {
+    pub const fn new(options: MqttOptions, topic: String) -> Result<Self, MqttError> {
+        Ok(Self { options, topic })
+    }
+
+    async fn connect(&self) -> Result<(AsyncClient, EventLoop), ClientError> {
+        let (client, eventloop) = AsyncClient::new(self.options.clone(), 1024);
+        client.subscribe(&self.topic, QoS::AtLeastOnce).await?;
+        Ok((client, eventloop))
+    }
+}
+
+pub struct MqttSource {
+    connector: MqttConnector,
+    decoder: Decoder,
+    log_namespace: LogNamespace,
+}
+
+impl MqttSource {
+    pub fn new(
+        connector: MqttConnector,
+        decoder: Decoder,
+        log_namespace: LogNamespace,
+    ) -> crate::Result<Self> {
+        Ok(Self {
+            connector,
+            decoder,
+            log_namespace,
+        })
+    }
+
+    pub async fn run(self, mut out: SourceSender, shutdown: ShutdownSignal) -> Result<(), ()> {
+        let (_client, mut connection) = self.connector.connect().await.map_err(|_| ())?;
+
+        loop {
+            tokio::select! {
+                _ = shutdown.clone() => return Ok(()),
+                mqtt_event = connection.poll() => {
+                    // If an error is returned here there is currently no way to tie this back
+                    // to the event that was posted which means we can't accurately provide
+                    // delivery guarantees.
+                    // We need this issue resolved first:
+                    // https://github.com/bytebeamio/rumqtt/issues/349
+                    match mqtt_event {
+                        Ok(MqttEvent::Incoming(Incoming::Publish(publish))) => {
+                            self.process_message(publish, &mut out).await;
+                        }
+                        Ok(MqttEvent::Incoming(
+                            Incoming::PubAck(_) | Incoming::PubRec(_) | Incoming::PubComp(_),
+                        )) => {
+                            // TODO Handle acknowledgement
+                        }
+                        _ => {}
+                    }
+                }
+            }
+        }
+    }
+
+    async fn process_message(&self, publish: Publish, out: &mut SourceSender) {
+        emit!(EndpointBytesReceived {
+            byte_size: publish.payload.len(),
+            protocol: "mqtt",
+            endpoint: &self.connector.options.broker_address().0,
+        });
+        let events_received = register!(EventsReceived);
```

Comment:
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏this should be saved as state in the struct, only registering it once on initialization.

Reply:
Just glancing at the `aws_sqs` source it looks to me like they're only cloning it once per the client_concurrency setting (which defaults to the number of threads on the system) , at initialization of the source.

Since the `mqtt` source isn't managing concurrency directly there shouldn't be a need to clone the EventsReceived registry.